packages/api/src/middleware/validation/checkProductValidity.js

```diff
@@ -18,39 +18,69 @@ import { handleObjectionError } from '../../util/errorCodes.js';
 import ProductModel from '../../models/productModel.js';
 import { checkAndTrimString } from '../../util/util.js';
 
+const taskProductRelationMap = {
+  soil_amendment_task: 'soil_amendment_product',
+  // pest_control_task: 'pest_control_product',
+  // cleaning_task: 'cleaning_product'
+};
+const productCheckMap = {
+  soil_amendment_task: checkSoilAmendmentProduct,
+  // pest_control_task: checkPestControlProduct,
+  // cleaning_task: checkCleaningProduct
+};
+
+function checkSoilAmendmentProduct(res, sap, isCreatingNew) {
+  const elements = [
+    'n',
+    'p',
+    'k',
+    'calcium',
+    'magnesium',
+    'sulfur',
+    'copper',
+    'manganese',
+    'boron',
+  ];
+  const molecularCompounds = ['ammonium', 'nitrate'];
+  // Check that element values are all positive
+  if (!elements.every((element) => !sap[element] || sap[element] >= 0)) {
+    return res.status(400).send('element values must all be positive');
+  }
+
+  // Check that a unit has been provided along with element values
+  if (isCreatingNew && elements.some((element) => sap[element]) && !sap.elemental_unit) {
```

Comment:
Is this different from the DB constraint `elemental_unit_check`?

Reply:
I think you are right this is a duplicate .. I just moved it here it isn't a new check.

But I will try removing that.